### PR TITLE
test(datetime): migrate tests to playwright

### PR DESCRIPTION
**Pull request by liamdebeasi**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue URL: N/A

## What is the new behavior?

- Sean mentioned that he would migrate the locale tests in https://github.com/ionic-team/ionic-framework/pull/25389, so I did not migrate them in mine.
- I removed the isDateEnabled test that checks to see if the console.error is logged as it was flaky across browsers (the listener was being added after the error was already logged). If you have alternative testing thoughts let me know!
- I removed the "values" tests as they were redundant.
- The `datetimeMonthDidChange` event is now dispatched on the window instead of the document. Our custom `spyOnEvent` only listens on the window at the moment. The effort to update this does not seem worth it at the moment.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by sean-perkins — COMMENTED** [review]
I believe all screenshots captured with "June 3" as the current date, will need to be modified to have a current value at least within the previous month, unless the test is overriding today's date reference.

Otherwise, screenshot tests will begin to fail as a result of:
1. Today's date will be different. So the circle around the day will be different between the captured screenshot and reference. 
2. Once the calendar month is within another month, the headings and calendar grid will not match the reference screenshot. 

Let me know if I am missing something with the tests 👍

**Comment by liamdebeasi** [comment]
Darn I thought I had caught all of those. Just updated -- should be good now.

File: core/src/components/datetime/test/display/datetime.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import { expect } from '@playwright/test';
+import { test } from '@utils/test/playwright';
+
+test.describe('datetime: display', () => {
+  test.describe('datetime: rendering', () => {
+    test('fixed size should not have visual regressions', async ({ page }) => {
```
Comment (amandaejohnston) [inline core/src/components/datetime/test/display/datetime.e2e.ts:6]: Some screenshots across a couple tests here are missing the solid circle highlighting today's value:
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-date-ios-ltr-Mobile-Safari-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-date-ios-rtl-Mobile-Safari-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-date-md-ltr-Mobile-Safari-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-date-md-rtl-Mobile-Safari-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-date-ios-ltr-Mobile-Safari-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-date-ios-rtl-Mobile-Safari-linux.png`

Comment (amandaejohnston) [inline core/src/components/datetime/test/display/datetime.e2e.ts:6]: A few time screenshots show the hour and minute columns scrolled up too far:
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-ios-ltr-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-ios-rtl-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-md-ltr-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-cover-time-md-rtl-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-time-ios-ltr-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-time-ios-rtl-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-time-md-ltr-Mobile-Firefox-linux.png`
- `core/src/components/datetime/test/display/datetime.e2e.ts-snapshots/datetime-display-time-md-rtl-Mobile-Firefox-linux.png`

File: core/src/components/datetime/test/values/index.html [hunk 2]
```diff
@@ -1,67 +0,0 @@
...
-        margin-left: 5px;
-      }
-
-      ion-datetime {
-        box-shadow: 0px 16px 32px rgba(0, 0, 0, 0.25), 0px 8px 16px rgba(0, 0, 0, 0.25);
-        border-radius: 8px;
-      }
-    </style>
-  </head>
-  <body>
```
Comment (amandaejohnston) [inline core/src/components/datetime/test/values/index.html:33]: What test was this redundant with? I don't see the `values` attributes tested anywhere else 🤔 Might just be missing it though.

Comment (liamdebeasi) [reply core/src/components/datetime/test/values/index.html:33]: Huh I think I may have gotten confused when removing this. I added an improved values test with Playwright.

File: core/src/components/datetime/test/display/datetime.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+      select.selectOption('date');
+      await page.waitForChanges();
+
+      expect(await page.screenshot()).toMatchSnapshot(`datetime-display-cover-date-${page.getSnapshotSettings()}.png`);
+    });
+  });
+  test.describe('datetime: switch presentations', () => {
+    test('month selection should work after changing presentation', async ({ page, browserName }) => {
+      test.skip(browserName === 'firefox', 'This test is flaky in Firefox');
```
[the remark this answers is not in the document]
Comment (liamdebeasi) [reply core/src/components/datetime/test/display/datetime.e2e.ts:86]: Ah that was left over from an old version of the test. I'll remove it

**Comment by amandaejohnston** [comment]
I think you'll need to set the values to be something in May or sooner, so the datetimes show that month instead of the current one. Otherwise, the circle around today's date will change from day to day and cause screenshot mismatches.

**Comment by liamdebeasi** [comment]
Fixed, thanks!

**Comment by liamdebeasi** [comment]
Ok now it should be good. When I re-ran screenshots, I did not delete the old screenshots first. There may not have been a high enough pixel difference for certain screenshots causing them to not get updated.

**Comment by amandaejohnston** [comment]
The screenshots for the footer tests are still showing June; maybe the screenshot job isn't looking at the right commit? 🤔 There might be other tests like that, I haven't looked.

**Comment by liamdebeasi** [comment]
🤦 I only deleted screenshots for the `color` test. Let me try this again...

**Review by amandaejohnston — APPROVED** [review]
Nice work!

**Review by sean-perkins — APPROVED** [review]
Nice work! 👍

**Pull request merged**